suppressPackageStartupMessages(library("gp"))
suppressPackageStartupMessages(library("argparse"))
parser = ArgumentParser(description="Genomic prediction cross-validation within and across populations. Please find the documentation within and interactive R session.")
parser$add_argument("-g", "--fname-geno",                                dest="fname_geno",                                      type="character",                                                                                  help="Filname of the input genotype data without missing data formatted as VCF with DP field and AD or GT field, RDS file containing an n samples x p loci allele frequency matrix with sample names in the rows and tab-delimited loci names in the column, and a TSV (tab-delimited) file of allele frequencies where the first 3 columns correspond to the chromosome or scaffold names, position, and allele which can be null, and the subsequent columns refer to the allele frequencies per sample where the column names are chr, pos, allele, followed by sample names without tabs.")
parser$add_argument("-p", "--fname-pheno",                               dest="fname_pheno",                                     type="character",                                                                                  help="Filename of the input phenotype data in text-delimited format, e.g. tab-delimited (tsv) and comma-separated values (csv). Contains at least 3 columns: sample IDs, population groupings, and trait values, where the sample IDs match the sample names in the genotype file.")
parser$add_argument("-n", "--population",                                dest="population",                                      type="character",                                                                                  help="Population ID found in the phenotype file where within population cross-validation will be performed on.")
parser$add_argument("-c", "--fname-covar",                               dest="fname_covar",                                     type="character", default=NULL,                                                                    help="Not yet implemented. Filename of the n samples x p variables covariate matrix where the row names match the sample names in the genotype and phenotype files [default=NULL].")
parser$add_argument("-o", "--dir-output",                                dest="dir_output",                                      type="character", default=NULL,                                                                    help="Output directory into which all temporary and final output files will be written [default=NULL].")
parser$add_argument("-s", "--geno-fname-snp-list",                       dest="geno_fname_snp_list",                             type="character", default=NULL,                                                                    help="Filename of the file containing the list of expected SNPs including their coordinates and alleles. This is a tab-delimited file with 3 columns: '#CHROM', 'POS', 'REF,ALT', corresponding to (Default=NULL) chromosome names (e.g. 'chr1' & 'chrom_1'),  numeric positions (e.g. 12345 & 100001), and  reference-alternative allele strings separated by a comma (e.g. 'A,T' & 'C,G' [default=NULL])")
parser$add_argument("-x", "--geno-ploidy",                               dest="geno_ploidy",                                     type="integer", default=NULL,                                                                      help="Expected ploidy level of the genotype file [default=NULL].")
parser$add_argument("--geno-bool-force-biallelic",                       dest="geno_bool_force_biallelic",                       type="logical", default=TRUE,                                                                      help="Force all loci including multi-allelic ones to be biallelic [default=TRUE]?")
parser$add_argument("--geno-bool-retain-minus-one-alleles-per-locus",    dest="geno_bool_retain_minus_one_alleles_per_locus",    type="logical", default=TRUE,                                                                      help="Remove the trailing allele per locus [default=TRUE]?")
parser$add_argument("--geno-min-depth",                                  dest="geno_min_depth",                                  type="integer", default=0,                                                                         help="If input is a VCF file: minimum depth per locus beyond which will be set to missing data [default=0].")
parser$add_argument("--geno-max-depth",                                  dest="geno_max_depth",                                  type="integer", default=.Machine$integer.max,                                                      help="If input is a VCF file: maximum depth per locus beyond which will be set to missing data [default=.Machine$integer.max].")
parser$add_argument("--geno-maf",                                        dest="geno_maf",                                        type="double", default=0.01,                                                                       help="Minimum allele frequency [default=0.01].")
parser$add_argument("--geno-sdev-min",                                   dest="geno_sdev_min",                                   type="double", default=0.0001,                                                                     help="Minimum allele frequency standard deviation [default=0.0001].")
parser$add_argument("--geno-max-n-alleles",                              dest="geno_max_n_alleles",                              type="integer", default=NULL,                                                                      help="Maximum number of alleles per locus [default=NULL].")
parser$add_argument("--geno-max-sparsity-per-locus",                     dest="geno_max_sparsity_per_locus",                     type="double", default=NULL,                                                                       help="Maximum sparsity or fraction of missing data per locus [default=NULL].")
parser$add_argument("--geno-frac-topmost-sparse-loci-to-remove",         dest="geno_frac_topmost_sparse_loci_to_remove",         type="double", default=NULL,                                                                       help="Fraction of the total number of loci from which the top most sparse loci will be removed [default=NULL].")
parser$add_argument("--geno-n-topmost-sparse-loci-to-remove",            dest="geno_n_topmost_sparse_loci_to_remove",            type="integer", default=NULL,                                                                      help="Number of top most sparse loci to be removed [default=NULL].")
parser$add_argument("--geno-max-sparsity-per-sample",                    dest="geno_max_sparsity_per_sample",                    type="double", default=NULL,                                                                       help="Maximum sparsity or fraction of missing data per sample [default=NULL].")
parser$add_argument("--geno-frac-topmost-sparse-samples-to-remove",      dest="geno_frac_topmost_sparse_samples_to_remove",      type="double", default=NULL,                                                                       help="Fraction of the total number of samples from which the top most sparse samples will be removed [default=NULL].")
parser$add_argument("--geno-n-topmost-sparse-samples-to-remove",         dest="geno_n_topmost_sparse_samples_to_remove",         type="integer", default=NULL,                                                                      help="Number of top most sparse samples to be removed [default=NULL].")
parser$add_argument("--pheno-sep",                                       dest="pheno_sep",                                       type="character", default="\t",                                                                    help="Delimiter used in the phenotype file [default='\t',].")
parser$add_argument("--pheno-header",                                    dest="pheno_header",                                    type="character", default=TRUE,                                                                    help="Does the phenotype file have a header line? [default=TRUE].")
parser$add_argument("--pheno-idx-col-id",                                dest="pheno_idx_col_id",                                type="integer", default=1,                                                                         help="Column number in the phenotype file corresponding to the sample names [default=1].")
parser$add_argument("--pheno-idx-col-pop",                               dest="pheno_idx_col_pop",                               type="integer", default=2,                                                                         help="Column number in the phenotype file corresponding to the population/grouping names [default=2].")
parser$add_argument("--pheno-idx-col-y",                                 dest="pheno_idx_col_y",                                 type="integer", default=3,                                                                         help="Column number in the phenotype file corresponding to the numeric phenotype data [default=3].")
parser$add_argument("--pheno-na-strings",                                dest="pheno_na_strings",                                type="character", default=c("", "-", "NA", "na", "NaN", "missing", "MISSING"),                     help="Strings of characters corresponding to missing data in the phenotype file [default=c('', '-', 'NA', 'na', 'NaN', 'missing', 'MISSING')].")
parser$add_argument("--pheno-bool-remove-outliers",                      dest="pheno_bool_remove_outliers",                      type="logical", default=TRUE,                                                                      help="Remove outliers from the phenotype file [default=TRUE]?")
parser$add_argument("--pheno-bool-remove-NA",                            dest="pheno_bool_remove_NA",                            type="logical", default=FALSE,                                                                     help="Remove samples missing phenotype data in the phenotype file? [default=FALSE].")
parser$add_argument("--bool-within",                                     dest="bool_within",                                     type="logical", default=TRUE,                                                                      help="Perform within population k-fold cross-validation? [default=TRUE].")
parser$add_argument("--bool-across",                                     dest="bool_across",                                     type="logical", default=FALSE,                                                                     help="Perform across populations cross-validations? [default=FALSE].")
parser$add_argument("--n-folds",                                         dest="n_folds",                                         type="integer", default=10,                                                                        help="Number of folds or training and validation sets for within population k-fold cross validation [default=10].")
parser$add_argument("--n-reps",                                          dest="n_reps",                                          type="integer", default=10,                                                                        help="Number of replications of random shuffling samples for each k-fold cross-validation in within population cross-validation [default=10].")
parser$add_argument("--vec-models-to-test",                              dest="vec_models_to_test",                              type="character", default=c("ridge","lasso","elastic_net","Bayes_A","Bayes_B","Bayes_C","gBLUP"),  help="Genomic prediction models to use [default=c('ridge','lasso','elastic_net','Bayes_A','Bayes_B','Bayes_C','gBLUP')].")
parser$add_argument("--bool-parallel",                                   dest="bool_parallel",                                   type="logical", default=TRUE,                                                                      help="Run the cross-validations in parallel [default=TRUE].")
parser$add_argument("--max-mem-Gb",                                      dest="max_mem_Gb",                                      type="double", default=15,                                                                         help="Maximum amount of memory available in gigabytes [default=15].")
parser$add_argument("--n-threads",                                       dest="n_threads",                                       type="integer", default=2,                                                                         help="Maximum number of computing threads available [default=2].")
parser$add_argument("--verbose",                                         dest="verbose",                                         type="logical", default=TRUE,                                                                      help="Show messages [default=TRUE]?")
args = parser$parse_args()
time_ini = Sys.time()
print("               ,@@@@@@@,")
print("       ,,,.   ,@@@@@@/@@,  .oo8888o.")
print("    ,&%%&%&&%,@@@@@/@@@@@@,8888|88/8o")
print("   ,%&|%&&%&&%,@@@|@@@/@@@88|88888/88'")
print("   %&&%&%&/%&&%@@|@@/ /@@@88888|88888'")
print("   %&&%/ %&%%&&@@| V /@@' `88|8 `/88'")
print("   `&%| ` /%&'    |.|        | '|8'")
print("       |o|        | |         | |")
print("       |.|        | |         | |")
print("    ||/ ._|//_/__/  ,|_//__||/.  |_//__/_")
print("Performing genomic prediction cross-validation using")
print(paste0("     - genotype file: ", args$fname_geno))
print(paste0("     - phenotype file: ", args$fname_pheno))
print(paste0("     - covariate file: ", args$fname_covar))
print(paste0("     - population: ", args$population))
print(paste0("     - within population CV: ", args$bool_within))
print(paste0("     - across populations CV: ", args$bool_across))
print(paste0("     - with a total of ", args$n_threads, " threads available and "))
print(paste0("       a total memory of ", args$max_mem_Gb, " Gb."))
print(paste0("Start time: ", time_ini))
print("Input parameters:")
print(args)
fname_out_Rds = gp::gp(args=args)
time_fin = Sys.time()
time_duration_minutes = as.numeric(difftime(time_fin, time_ini, units="min"))
print("-----------------------------------------------------------")
print("-----------------------------------------------------------")
print("-----------------------------------------------------------")
print(paste0("End time: ", time_fin))
print(paste0(" Finished after ", time_duration_minutes, " minutes"))
if (methods::is(fname_out_Rds, "gpError")) {
    print("ERROR:")
    print(fname_out_Rds)
} else {
    print(paste0(" Output Rds file: ", fname_out_Rds))
}
print("        |   ^__^")
print("         |_ (oo)|_______")
print("            (__)|       )|/|")
print("                ||----w |")
print("                ||     ||")
